scipy/optimize/tnc/_tnc_module.pyx

```diff
@@ -0,0 +1,205 @@
+
+ctypedef np.float64_t float64_t
+
+
+cdef extern from "tnc.h":
+    ctypedef void tnc_callback(double[], void*) except *
+    ctypedef int tnc_function(double[], double*, double[], void*) except 1
+
+    cdef enum tnc_rc:
+        TNC_MINRC
+        TNC_ENOMEM
+        TNC_EINVAL
+        TNC_INFEASIBLE
+        TNC_LOCALMINIMUM
+        TNC_FCONVERGED
+        TNC_XCONVERGED
+        TNC_MAXFUN
+        TNC_LSFAIL
+        TNC_CONSTANT
+        TNC_NOPROGRESS
+        TNC_USERABORT
+
+    cdef int tnc(int n, double x[], double *f, double g[],
+            tnc_function *function, void *state,
+            double low[], double up[], double scale[], double offset[],
+            int messages, int maxCGit, int maxnfeval, double eta, double stepmx,
+            double accuracy, double fmin, double ftol, double xtol, double pgtol,
+            double rescale, int *nfeval, int *niter, tnc_callback *tnc_callback) except *
+
+
+cdef struct s_pytnc_state:
+  void *py_function
+  void *py_callback
+  int n
+  int failed
+ctypedef s_pytnc_state pytnc_state
+
+
+cdef int function(double x[], double *f, double g[], void *state) except 1:
+    # return 0 if no error, 1 if error
+    # However, if the user function raises Exceptions Cython should deal
+    # with it.
+
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+        double *g_data
+
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    xcopy = np.empty(n, dtype=float)
+    x_data = <float64_t *>np.PyArray_DATA(xcopy)
+    memcpy(x_data, x, sizeof(double) * n)
+
+    fx, gx = (<object>py_state.py_function)(xcopy)
+
+    if not np.isscalar(fx):
+        try:
+            fx = np.asarray(fx).item()
+        except (TypeError, ValueError) as e:
+            raise ValueError(
+                "The user-provided objective function "
+                "must return a scalar value."
+            ) from e
+
+    f[0] = <double> fx
+
+    gx = np.asfarray(gx)
+    if gx.size != n:
+        raise ValueError("tnc: gradient must have shape (len(x0),)")
+
+    if not gx.flags['C_CONTIGUOUS']:
+        gx = np.ascontiguousarray(gx, dtype=np.float64)
+
+    g_data = <float64_t *>np.PyArray_DATA(gx)
+    memcpy(g, g_data, n * sizeof(double))
+
+    return 0
+
+
+cdef void callback_function(double x[], void *state) except *:
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    try:
+        xcopy = np.empty(n, dtype=float)
+        x_data = <float64_t *>np.PyArray_DATA(xcopy)
+        memcpy(x_data, x, sizeof(double) * n)
+    except MemoryError as exc:
+        # couldn't allocate memory
+        py_state.failed = 1
+        raise exc
+
+    val = (<object>py_state.py_callback)(xcopy)
```

Comment:
This can also raise an exception I imagine - should that set `py_state.failed = 1` too?

Reply:
At this point the callstack is:
```
Python --> Cython ([L197](https://github.com/scipy/scipy/pull/14882/files#diff-f29933549abd450f6377e8daef0e671508e127b1b92db3e6c9c3b376bda0c2c6R197)) --> C ([tnc.c L804](https://github.com/scipy/scipy/blob/master/scipy/optimize/tnc/tnc.c#L804)) --> Cython (L114) --> Python (callback)
```

I don't know how Cython works here. If the Python callback raises an exception, then how does the Cython code here propagate the exception? Specifically, will the memory cleanup at the end of the `tnc` function ([tnc.c, L909](https://github.com/scipy/scipy/blob/master/scipy/optimize/tnc/tnc.c#L909)) ever run? If it doesn't then would there be a memory leak?